apps/prairielearn/elements/pl-big-o-input/pl-big-o-input.py

```diff
@@ -59,6 +59,8 @@ def prepare(element_html: str, data: pl.QuestionData) -> None:
     pl.check_attribs(element, required_attribs, optional_attribs)
 
     name = pl.get_string_attrib(element, "answers-name")
+    pl.check_answers_names(data, name)
+
```

Comment:
I don't like that we have to call this function manually from every single element. Isn't there some way we can have it just executed automatically before the `prepare` function of each function? or even implement it in the Javascript?

Reply:
Something like this might be possible in the experimental renderer. But right now, there's no way to collect this single attribute value across all elements with the current separation of logic. This goes back to a structural issue about elements in general, the Python backends are really just a collection of functions and it's hard to share functionality between them (which is why all the graded elements have a bunch of similar-looking boilerplate code). Even the use of `answers-name` as an attribute is by convention.

Reply:
I have the same reservation about this change. As it stands, I had to manually check every element, and this PR indeed neglects to make the appropriate change for some of them. At the same time, as Eliot points out, this name is just a convention. An element could in theory use `answers-name` to identify a specific key to read from (not write to), which would be completely valid *and* fail any kind of automatic checks we build into the renderer.

This concern shouldn't _necessarily_ stand in the way of merging this change, but I'm also not going to rubber-stamp it without a bit more thought/conversation.

Reply:
I think that's perfectly reasonable. As you said Nate, the reason I'm in support of a change like this is the lack of a unified system for dealing with element attributes. The ideal thing to do would be to have a dataclass-like interface for the developer, where they could specify the attributes on each element (with appropriate defaults) and define some type of validation logic on top of all of this, instead of doing everything manually the way it is now. Then this change could be integrated in the background as part of that validation logic.

However, until something like this gets written, we don't really have a choice and any addition of validation / consistency logic has to be done manually. Even bolting this on top of something like `check_attribs` could cause unexpected issues with people's client code.